Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift

```diff
@@ -48,16 +49,49 @@ public extension SwiftSyntaxRule {
         }
 
         return visitor
-            .walk(file: file, handler: \.violationPositions)
+            .walk(file: file, handler: \.violations)
             .sorted()
-            .map { makeViolation(file: file, position: $0) }
+            .map { makeViolation(file: file, violation: $0) }
+    }
+}
+
+/// A violation produced by `ViolationsSyntaxVisitor`s.
+public struct ReasonedRuleViolation: Comparable {
+    /// The violation's position.
+    public let position: AbsolutePosition
+    /// A specific reason for the violation.
+    public let reason: String?
```

Comment:
while we're at this, can we add the `severity`? this is the other reason rules can't use `ViolationsSyntaxVisitor` currently

Reply:
I already had it in there but cannot remember why I decided to leave it out at the end.

Could you mention an example rule for which this would be useful?

Reply:
any of the "Count" violations, like `FunctionParameterCountRule`